In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

In [2]:
df_a = pd.read_csv("/home/spaceape/bachelor_thesis/data/lending-club/sample_acc.csv")
df_r = pd.read_csv("/home/spaceape/bachelor_thesis/data/lending-club/sample_rej.csv")

#df_a = pd.read_csv("/home/spaceape/bachelor_thesis/data/lending-club/accepted_2007_to_2017.csv")
#df_r = pd.read_csv("/home/spaceape/bachelor_thesis/data/lending-club/sample_rej.csv")


# 1. DATA WRANGLING
#count number of nans within each feature
nans_fa = pd.Series(df_a.isnull().sum(axis=0), name="nans")
nans_fa = nans_fa.sort_values(ascending=False)
nans_oa = pd.Series(df_a.isnull().sum(axis=1), name="nans")
nans_oa = nans_oa.sort_values(ascending=False)

/home/spaceape/anaconda3/envs/keras/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (1,123,124,125,128,129,130,133) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
len(df_a)

200000

In [4]:
# outcome variable
df_a["loan_status"].head()

0     Fully Paid
1     Fully Paid
2    Charged Off
3    Charged Off
4     Fully Paid
Name: loan_status, dtype: object

In [5]:
df_a.columns

Index(['Unnamed: 0', 'id', 'member_id', 'loan_amnt', 'funded_amnt',
       'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade',
       ...
       'hardship_amount', 'hardship_start_date', 'hardship_end_date',
       'payment_plan_start_date', 'hardship_length', 'hardship_dpd',
       'hardship_loan_status', 'orig_projected_additional_accrued_interest',
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount'],
      dtype='object', length=137)

In [6]:
df_r.columns

Index(['Unnamed: 0', 'Amount Requested', 'Application Date', 'Loan Title',
       'Risk_Score', 'Debt-To-Income Ratio', 'Zip Code', 'State',
       'Employment Length', 'Policy Code'],
      dtype='object')

In [7]:
df_r.head()

Unnamed: 0  Amount Requested Application Date          Loan Title  \
0    13397118            4000.0       2015-10-19  debt_consolidation   
1     5639475           35000.0       2014-10-04  debt_consolidation   
2    12797198           12000.0       2015-08-06  debt_consolidation   
3     2167937            8000.0       2016-10-11  debt_consolidation   
4     9826493           15000.0       2017-06-28    Home improvement   

   Risk_Score Debt-To-Income Ratio Zip Code State Employment Length  \
0         NaN               13.42%    170xx    PA          < 1 year   
1       687.0               30.48%    018xx    MA          < 1 year   
2         NaN               28.69%    346xx    FL          < 1 year   
3         NaN               38.48%    750xx    TX          < 1 year   
4       676.0                5.91%    014xx    MA          < 1 year   

   Policy Code  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0

In [8]:
for i in df_r.columns:
    print(df_r[i].value_counts())

7715386     1
10308348    1
11293407    1
12561970    1
8795457     1
7975651     1
12552932    1
105190      1
12375311    1
6421226     1
3476604     1
931155      1
3261165     1
6806254     1
10343151    1
6497009     1
4797171     1
8975541     1
1162856     1
13993427    1
6886135     1
13722361    1
11093374    1
5917404     1
7502554     1
287449      1
11598538    1
6927351     1
13937344    1
5832844     1
           ..
5844431     1
1037776     1
6409683     1
9960917     1
1283544     1
13800888    1
7660982     1
3900826     1
7890342     1
13594122    1
2733468     1
8889758     1
7624096     1
9737255     1
10634658    1
1199523     1
10384804    1
9661444     1
2305912     1
6487477     1
3282344     1
3278250     1
5508523     1
1320366     1
2881222     1
1994887     1
10155599    1
13026738    1
3344042     1
234412      1
Name: Unnamed: 0, Length: 100000, dtype: int64
10000.0     11786
5000.0      11134
20000.0      7550
15000.0      7084
1000.0       6695
35000.0  

In [9]:
for col in df_r.columns:
    print(col in df_a.columns)

True
False
False
False
False
False
False
False
False
False


In [10]:
# drop the features containing more than 10% of NaNs
for feature in df_a[:]:
    if (nans_fa[feature] > (0.2*len(df_a))):
        print (feature, nans_fa[feature])
        df_a = df_a.drop(feature, 1)

id 199998
member_id 200000
desc 183389
mths_since_last_delinq 99360
mths_since_last_record 165860
next_pymnt_d 87314
mths_since_last_major_derog 147027
annual_inc_joint 197272
dti_joint 197272
verification_status_joint 197272
open_acc_6m 114062
open_il_6m 114061
open_il_12m 114061
open_il_24m 114061
mths_since_rcnt_il 116328
total_bal_il 114061
il_util 125372
open_rv_12m 114061
open_rv_24m 114061
max_bal_bc 114061
all_util 114069
inq_fi 114061
total_cu_tl 114062
inq_last_12m 114062
mths_since_recent_bc_dlq 151997
mths_since_recent_revol_delinq 132004
revol_bal_joint 198974
sec_app_earliest_cr_line 198974
sec_app_inq_last_6mths 198974
sec_app_mort_acc 198974
sec_app_open_acc 198974
sec_app_revol_util 198987
sec_app_open_il_6m 198974
sec_app_num_rev_accts 198974
sec_app_chargeoff_within_12_mths 198974
sec_app_collections_12_mths_ex_med 198974
sec_app_mths_since_last_major_derog 199637
hardship_type 199862
hardship_reason 199862
hardship_status 199862
deferral_term 199862
hardship_amount 

In [11]:
# print number of missing observation of given feature
for feature in df_a[:]:
    if (nans_fa[feature] > 0):
        print (feature, nans_fa[feature])

loan_amnt 2
funded_amnt 2
funded_amnt_inv 2
term 2
int_rate 2
installment 2
grade 2
sub_grade 2
emp_title 12187
emp_length 11309
home_ownership 2
annual_inc 2
verification_status 2
issue_d 2
loan_status 2
pymnt_plan 2
purpose 2
title 2999
zip_code 3
addr_state 2
dti 2
delinq_2yrs 6
earliest_cr_line 6
inq_last_6mths 7
open_acc 6
pub_rec 6
revol_bal 2
revol_util 112
total_acc 6
initial_list_status 2
out_prncp 2
out_prncp_inv 2
total_pymnt 2
total_pymnt_inv 2
total_rec_prncp 2
total_rec_int 2
total_rec_late_fee 2
recoveries 2
collection_recovery_fee 2
last_pymnt_d 1262
last_pymnt_amnt 2
last_credit_pull_d 10
collections_12_mths_ex_med 24
policy_code 2
application_type 2
acc_now_delinq 6
tot_coll_amt 9242
tot_cur_bal 9242
total_rev_hi_lim 9242
acc_open_past_24mths 6529
avg_cur_bal 9244
bc_open_to_buy 8440
bc_util 8539
chargeoff_within_12_mths 24
delinq_amnt 6
mo_sin_old_il_acct 14933
mo_sin_old_rev_tl_op 9242
mo_sin_rcnt_rev_tl_op 9242
mo_sin_rcnt_tl 9242
mort_acc 6529
mths_since_recent_bc

In [12]:
df_a = df_a.drop(["Unnamed: 0"], 1)
#df_a = df_a.drop(["Unnamed: 0", "hardship_flag", "emp_title", "zip_code", "last_pymnt_d", "last_credit_pull_d", "issue_d", "earliest_cr_line", "title"], 1)
#df_a = df_a.drop(["hardship_flag", "emp_title", "zip_code", "last_pymnt_d", "last_credit_pull_d", "issue_d", "earliest_cr_line", "title"], 1)

In [13]:
df_a.shape

(200000, 85)

In [14]:
# missing number of features
nans_oa = pd.Series(df_a.isnull().sum(axis=1), name="nans")
nans_oa = nans_oa.sort_values(ascending=False)
nans_oa.value_counts()

0     145031
1      30873
2      10101
34      5917
3       2631
4       1583
23      1291
25       953
35       398
5        352
26       193
24       193
36       190
6        141
27        54
7         41
28        16
38        12
29         9
8          5
30         4
37         4
39         2
47         2
85         2
49         1
48         1
Name: nans, dtype: int64

In [15]:
# delete observations with x or more missing features
drp = nans_oa > 30
drp = drp[drp==True].index
df_a = df_a.drop(drp, axis=0)

nans_fa = pd.Series(df_a.isnull().sum(axis=0), name="nans")
nans_fa = nans_fa.sort_values(ascending=False)
nans_oa = pd.Series(df_a.isnull().sum(axis=1), name="nans")
nans_oa = nans_oa.sort_values(ascending=False)


In [16]:
len(df_a)

193471

In [17]:
df_a["loan_status"].value_counts()

Current               106512
Fully Paid             64464
Charged Off            16667
Late (31-120 days)      2915
In Grace Period         1882
Late (16-30 days)        672
Default                  269
Issued                    90
Name: loan_status, dtype: int64

In [18]:
df_a["emp_length"] = df_a["emp_length"].replace(["< 1 year", "1 year", "2 years", "3 years", "4 years", "5 years", "6 years", "7 years", "8 years", "9 years", "10 years", "10+ years"], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15])

In [19]:
df_a.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0    10925.0      10925.0          10925.0   36 months     15.59       381.89   
1     8500.0       8500.0           8500.0   36 months     14.09       290.89   
2    12500.0      12500.0          12500.0   36 months     13.35       423.29   
3     7200.0       7200.0           7200.0   36 months     12.69       241.53   
4    12000.0      12000.0          12000.0   36 months     10.99       392.81   

  grade sub_grade             emp_title  emp_length      ...       \
0     C        C5           Planner II          0.0      ...        
1     B        B5  HEI Hotels & Resorts         3.0      ...        
2     C        C2    Equipment Operator         0.0      ...        
3     C        C2                   NaN         NaN      ...        
4     B        B4     Senior Accountant         4.0      ...        

  num_tl_op_past_12m  pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies  \
0                1.0           100.0              0.0                  0.0   
1                1.0           100.0             66.7                  1.0   
2                0.0           100.0             50.0                  0.0   
3                5.0           100.0              0.0                  0.0   
4                4.0           100.0             40.0                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0         50284.0           30094.0         3500.0   
1       0.0         40091.0           24719.0        11700.0   
2       0.0         98571.0           25488.0         4300.0   
3       0.0        103900.0            2111.0         3000.0   
4       0.0        201040.0           58218.0        22500.0   

  total_il_high_credit_limit hardship_flag  
0                    37984.0             N  
1                    28091.0             N  
2                    26649.0             N  
3                        0.0             N  
4                    56242.0             N  

[5 rows x 85 columns]

In [20]:
# categorical variables
factors = []
for feat in df_a:
    if (type(df_a[feat][0]) == type("string") and feat != "loan_status"):
        factors.append(feat)
        print (feat, nans_fa[feat])

term 0
grade 0
sub_grade 0
emp_title 11779
home_ownership 0
verification_status 0
issue_d 0
pymnt_plan 0
purpose 0
title 2995
zip_code 1
addr_state 0
earliest_cr_line 0
initial_list_status 0
last_pymnt_d 1245
last_credit_pull_d 7
application_type 0
hardship_flag 0


In [21]:
#numerical variables
for feat in df_a:
    if (type(df_a[feat][0]) != type("string") and feat != "loan_status"): 
        print (feat, nans_fa[feat])

loan_amnt 0
funded_amnt 0
funded_amnt_inv 0
int_rate 0
installment 0
emp_length 11118
annual_inc 0
dti 0
delinq_2yrs 0
inq_last_6mths 1
open_acc 0
pub_rec 0
revol_bal 0
revol_util 97
total_acc 0
out_prncp 0
out_prncp_inv 0
total_pymnt 0
total_pymnt_inv 0
total_rec_prncp 0
total_rec_int 0
total_rec_late_fee 0
recoveries 0
collection_recovery_fee 0
last_pymnt_amnt 0
collections_12_mths_ex_med 0
policy_code 0
acc_now_delinq 0
tot_coll_amt 2713
tot_cur_bal 2713
total_rev_hi_lim 2713
acc_open_past_24mths 0
avg_cur_bal 2715
bc_open_to_buy 1911
bc_util 2010
chargeoff_within_12_mths 0
delinq_amnt 0
mo_sin_old_il_acct 8404
mo_sin_old_rev_tl_op 2713
mo_sin_rcnt_rev_tl_op 2713
mo_sin_rcnt_tl 2713
mort_acc 0
mths_since_recent_bc 1787
mths_since_recent_inq 20540
num_accts_ever_120_pd 2713
num_actv_bc_tl 2713
num_actv_rev_tl 2713
num_bc_sats 1156
num_bc_tl 2713
num_il_tl 2713
num_op_rev_tl 2713
num_rev_accts 2713
num_rev_tl_bal_gt_0 2713
num_sats 1156
num_tl_120dpd_2m 10518
num_tl_30dpd 2713
num_tl_

In [22]:
df_a = df_a.reindex(range(0, len(df_a)))

In [23]:
df_a

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  \
0         10925.0      10925.0          10925.0   36 months     15.59   
1          8500.0       8500.0           8500.0   36 months     14.09   
2         12500.0      12500.0          12500.0   36 months     13.35   
3          7200.0       7200.0           7200.0   36 months     12.69   
4         12000.0      12000.0          12000.0   36 months     10.99   
5          8000.0       8000.0           8000.0   36 months      9.49   
6         15325.0      15325.0          15325.0   60 months     18.25   
7          3475.0       3475.0           3475.0   36 months     30.84   
8          3750.0       3750.0           3750.0   36 months     11.99   
9         20000.0      20000.0          20000.0   36 months      5.32   
10        15000.0      15000.0          15000.0   36 months      6.89   
11        16800.0      16800.0          16800.0   36 months      6.49   
12         5000.0       5000.0           5000.0   36 months      8.67   
13        24000.0      24000.0          24000.0   36 months      7.89   
14        18975.0      18975.0          18975.0   60 months      9.99   
15        16000.0      16000.0          16000.0   36 months      6.24   
16        14000.0      14000.0          14000.0   36 months      7.99   
17         6400.0       6400.0           6400.0   36 months     12.35   
18         3600.0       3600.0           3600.0   36 months     14.65   
19        10000.0      10000.0          10000.0   36 months      5.32   
20        13000.0      13000.0          13000.0   36 months     13.35   
21        20000.0      20000.0          19950.0   36 months     10.64   
22         2000.0       2000.0           2000.0   36 months     14.48   
23         5000.0       5000.0           5000.0   36 months      8.49   
24         1500.0       1500.0           1500.0   36 months     16.55   
25         6800.0       6800.0           6800.0   36 months     11.49   
26        14000.0      14000.0          14000.0   36 months     16.29   
27         5000.0       5000.0           5000.0   36 months     13.67   
28        11500.0      11500.0          11500.0   36 months     17.86   
29        22800.0      22800.0          22800.0   36 months     18.25   
...           ...          ...              ...         ...       ...   
193441    10775.0      10775.0          10775.0   60 months     24.08   
193442    15000.0      15000.0          15000.0   60 months     12.74   
193443     5000.0       5000.0           5000.0   36 months     13.99   
193444    35000.0      35000.0          34975.0   60 months     20.49   
193445    36550.0      36550.0          36550.0   60 months      7.89   
193446    28000.0      28000.0          28000.0   60 months      8.18   
193447     3200.0       3200.0           3200.0   36 months      8.18   
193448    16000.0      16000.0          16000.0   60 months      9.99   
193449     6000.0       6000.0           6000.0   36 months     14.46   
193450     6400.0       6400.0           6400.0   36 months     13.98   
193451     6025.0       6025.0           6025.0   36 months     11.49   
193452    35000.0      35000.0          35000.0   60 months     25.44   
193453    14000.0      14000.0          14000.0   36 months      8.59   
193454     6500.0       6500.0           6500.0   36 months      9.80   
193455    35000.0      35000.0          35000.0   60 months     10.99   
193456    28000.0      28000.0          28000.0   36 months      7.90   
193457    17550.0      17550.0          17500.0   36 months     24.74   
193458    14000.0      14000.0          14000.0   60 months     12.39   
193459    30000.0      30000.0          29975.0   36 months     11.39   
193460     5000.0       5000.0           5000.0   36 months      5.32   
193461     3625.0       3625.0           3625.0   36 months     14.46   
193462     6000.0       6000.0           6000.0   36 months     14.85   
193463    30000.0      30000.0          30000.0   36 months     18

In [24]:
len(df_a)

193471

In [25]:
# from sklearn.preprocessing import Imputer
# imp_str = Imputer(missing_values="NaN", strategy='most_frequent', axis=0)
# imp_num = Imputer(missing_values="NaN", strategy='median', axis=0)

# for feat in df_a.columns:
#     if (nans_fa[feat] > 0) and (feat != "loan_status"):
#         if feat in factors:
#             #df_a[feat] = imp_str.fit_transform(np.reshape(df_a[feat], (len(df_a[feat]), 1)))
#             tmp = df_a[feat].values.reshape((len(df_a[feat])))
#             print (tmp)
#             #print (imp_str.fit_transform(tmp), 1)
#         else: 
#             df_a[feat] = imp_num.fit_transform(np.reshape(df_a[feat], (len(df_a[feat]), 1)))
#             #tmp = df_a[feat].values.reshape(len(df_a[feat]))
#             #print (tmp)
#             #print (imp_num.fit_transform(tmp, 1))
            
# #df_a[factors] = pd.DataFrame(imp_str.fit_transform(df_a[factors]), index=df_a.index, columns=df_a[factors].columns)

In [26]:
nans_fa = pd.Series(df_a.isnull().sum(axis=0), name="nans")
nans_fa = nans_fa.sort_values(ascending=False)
nans_oa = pd.Series(df_a.isnull().sum(axis=1), name="nans")
nans_oa = nans_oa.sort_values(ascending=False)

In [27]:
# manual encoding dependent variable

stat_used = ["Current", "Fully Paid", "Default", "Charged Off"]
default = [0, 0, 1, 1]
stat_notused = []
nans = []

for i in (df_a["loan_status"].unique()):
    if i not in stat_used:
        stat_notused.append(i)
        nans.append(np.nan)

#df_a["loan_status"] = df_a["loan_status"].replace(["Late (31-120 days)", "Late (16-30 days)", "In Grace Period", "Issued", "Does not meet the credit policy. Status:Fully Paid", "Does not meet the credit policy. Status:Charged Off"],[np.nan, np.nan, np.nan, np.nan, np.nan, np.nan])
df_a["loan_status"] = df_a["loan_status"].replace(stat_notused, nans)
drp = df_a["loan_status"] == np.nan
drp = drp[drp==True].index
df_a = df_a.drop(drp, axis=0)
df_a["loan_status"] = df_a["loan_status"].replace(stat_used, default)

In [28]:
df_a = df_a.dropna()

In [29]:
df_a["loan_status"].value_counts()

0.0    123866
1.0     12368
Name: loan_status, dtype: int64

In [30]:
df_a.shape

(136234, 85)

In [31]:
df_a.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0    10925.0      10925.0          10925.0   36 months     15.59       381.89   
1     8500.0       8500.0           8500.0   36 months     14.09       290.89   
2    12500.0      12500.0          12500.0   36 months     13.35       423.29   
4    12000.0      12000.0          12000.0   36 months     10.99       392.81   
5     8000.0       8000.0           8000.0   36 months      9.49       256.23   

  grade sub_grade               emp_title  emp_length      ...       \
0     C        C5             Planner II          0.0      ...        
1     B        B5    HEI Hotels & Resorts         3.0      ...        
2     C        C2      Equipment Operator         0.0      ...        
4     B        B4       Senior Accountant         4.0      ...        
5     B        B2  Database Administrator         2.0      ...        

  num_tl_op_past_12m  pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies  \
0                1.0           100.0              0.0                  0.0   
1                1.0           100.0             66.7                  1.0   
2                0.0           100.0             50.0                  0.0   
4                4.0           100.0             40.0                  0.0   
5                0.0           100.0             33.3                  0.0   

   tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0        0.0         50284.0           30094.0         3500.0   
1        0.0         40091.0           24719.0        11700.0   
2        0.0         98571.0           25488.0         4300.0   
4        0.0        201040.0           58218.0        22500.0   
5        0.0         39794.0           26704.0         7400.0   

  total_il_high_credit_limit hardship_flag  
0                    37984.0             N  
1                    28091.0             N  
2                    26649.0             N  
4                    56242.0             N  
5                    32394.0             N  

[5 rows x 85 columns]

In [32]:
y = df_a["loan_status"]
y = np.asarray(y, dtype=np.int)
df_a = df_a.drop(["loan_status"], 1)

In [33]:
#df_a = df_a.drop(["emp_title"], 1)

# remove categorical vars with only one output or more than 5000 output values (emp_title)
for feat in factors:
    if len(df_a[feat].unique()) == 1 or len(df_a[feat].unique()) > 200:
        df_a = df_a.drop([feat], 1)

# categorical variables
factors = list()
factors_idx = list()
for feat in df_a:
    if (type(df_a[feat][0]) == type("string")):
        factors.append(feat)
        factors_idx.append(df_a.columns.get_loc(feat))
        print (feat, len(df_a[feat].unique()))
        

term 2
grade 7
sub_grade 35
home_ownership 6
verification_status 3
issue_d 59
purpose 13
addr_state 50
initial_list_status 2
last_pymnt_d 59
last_credit_pull_d 59
application_type 3


In [34]:
factors_idx

[3, 6, 7, 9, 11, 12, 13, 14, 23, 33, 35, 38]

In [35]:
#numerical variables
num = list()
num_idx = list()
for feat in df_a:
    if (type(df_a[feat][0]) != type("string") and feat != "loan_status"):
        num.append(feat)
        num_idx.append(df_a.columns.get_loc(feat))
        #print (feat, len(df_a[feat].unique()))

In [36]:
print(num_idx)

[0, 1, 2, 4, 5, 8, 10, 15, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 36, 37, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77]


In [37]:
# label encoding categorical variables
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

for feat in factors:
    #print (feat, len(df_a[feat].unique()), df_a[feat].unique())
    a = le.fit_transform(df_a[feat])
    df_a[feat] = a

In [38]:
# one-hot-encoding categorical variables
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse=False)
X_cat = []

for feat in factors:
    a = enc.fit_transform(df_a[feat].reshape(len(df_a), 1))
    X_cat.append(a)
    #print (a)

# for feat in factors:
#     a = df_a[feat].reshape((len(df_a),1))
#     a = enc.fit_transform(a)
#     #print(a.shape)
#     #X_cat[:,factors.index(feat)] = a
#     np.insert(X_cat, a.shape[1], a)
#     print (type(a))

/home/spaceape/anaconda3/envs/keras/lib/python3.6/site-packages/ipykernel_launcher.py:7: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  import sys


In [39]:
df_a.shape

(136234, 78)

In [40]:
# preparation for embedding
cat_sz = [(c, len(df_a[c].unique())+1) for c in factors]
cat_sz

[('term', 3),
 ('grade', 8),
 ('sub_grade', 36),
 ('home_ownership', 7),
 ('verification_status', 4),
 ('issue_d', 60),
 ('purpose', 14),
 ('addr_state', 51),
 ('initial_list_status', 3),
 ('last_pymnt_d', 60),
 ('last_credit_pull_d', 60),
 ('application_type', 4)]

In [41]:
emb_szs = [(c, min(50, (c+1)//2)) for _,c in cat_sz]
emb_szs

[(3, 2),
 (8, 4),
 (36, 18),
 (7, 4),
 (4, 2),
 (60, 30),
 (14, 7),
 (51, 26),
 (3, 2),
 (60, 30),
 (60, 30),
 (4, 2)]

In [42]:
# might be too rough to drop all the nans observations
df_a = df_a.dropna()
len(df_a)

136234

In [43]:
df_a = df_a.reset_index(drop=True)
df_a.head()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  \
0    10925.0      10925.0          10925.0     0     15.59       381.89   
1     8500.0       8500.0           8500.0     0     14.09       290.89   
2    12500.0      12500.0          12500.0     0     13.35       423.29   
3    12000.0      12000.0          12000.0     0     10.99       392.81   
4     8000.0       8000.0           8000.0     0      9.49       256.23   

   grade  sub_grade  emp_length  home_ownership             ...              \
0      2         14         0.0               5             ...               
1      1          9         3.0               5             ...               
2      2         11         0.0               1             ...               
3      1          8         4.0               1             ...               
4      1          6         2.0               5             ...               

   num_tl_90g_dpd_24m  num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  \
0                 0.0                 1.0           100.0               0.0   
1                 0.0                 1.0           100.0              66.7   
2                 0.0                 0.0           100.0              50.0   
3                 0.0                 4.0           100.0              40.0   
4                 0.0                 0.0           100.0              33.3   

   pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
0                   0.0        0.0          50284.0            30094.0   
1                   1.0        0.0          40091.0            24719.0   
2                   0.0        0.0          98571.0            25488.0   
3                   0.0        0.0         201040.0            58218.0   
4                   0.0        0.0          39794.0            26704.0   

   total_bc_limit  total_il_high_credit_limit  
0          3500.0                     37984.0  
1         11700.0                     28091.0  
2          4300.0                     26649.0  
3         22500.0                     56242.0  
4          7400.0                     32394.0  

[5 rows x 78 columns]

In [44]:
# setting up the structure of dataset
X = np.asarray(df_a)
X_num = np.asarray(df_a[num])

features = df_a.columns

In [45]:
#print(np.hstack(X_cat))
print(np.concatenate(X_cat, axis=1).shape)

(136234, 298)


In [46]:
# cross validation
from sklearn.model_selection import StratifiedKFold
skf = StratifiedKFold(n_splits=8, shuffle=True, random_state = 12)

for train_index, test_index in skf.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]    
#     X_num_train, x_num_test = X_num[train_index], X_num[test_index]

print("%s train %s test" % (len(y_train), len(y_test)))

X_cat_train = []
X_cat_test = []

for i in range(0, len(X_cat)):
    X_cat_train.append(X_cat[i][train_index])
    X_cat_test.append(X_cat[i][test_index])

119205 train 17029 test


In [47]:
#print("%s min %s max" % (np.min(X_train[:,0]), np.max(X_train[:,0])))

In [48]:
# feature scaling
# X_train, X_test contain scaled numerical values and labeled categorical
from sklearn import preprocessing
mascaler = preprocessing.MaxAbsScaler()
X_train[:, num_idx] = mascaler.fit_transform(X_train[:, num_idx])
X_test[:, num_idx] = mascaler.fit_transform(X_test[:, num_idx])

In [49]:
# concatenate numerical arrays and categorical arrays (dummy variables)
# XX_train, XX_test contain scaled numerical values and one-hot-encoded categorical
XX_train = np.concatenate(X_cat_train, axis=1)
XX_train = np.concatenate((XX_train, X_train[:, num_idx]), axis=1)

XX_test = np.concatenate(X_cat_test, axis=1)
XX_test = np.concatenate((XX_test, X_test[:, num_idx]), axis=1)

In [50]:
#random seeds for stochastic parts of neural network 
np.random.seed(10)
from tensorflow import set_random_seed
set_random_seed(15)

from keras.models import Sequential
from keras.layers import Dense, Activation, Merge, Reshape, Dropout
from keras.layers.embeddings import Embedding

/home/spaceape/anaconda3/envs/keras/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using Theano backend.
WARNING (theano.configdefaults): install mkl with `conda install mkl-service`: No module named 'mkl'


In [51]:
# model of NN composition
model = Sequential()
model.add(Dense(len(XX_train), input_dim=XX_train.shape[1], activation="relu"))
model.add(Dropout(0.3))
model.add(Dense(15, activation="relu"))
model.add(Dense(1, activation="sigmoid"))

model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=["accuracy"])

In [53]:
emb = model.fit(XX_train, y_train, validation_split=0.25, epochs=1, batch_size=1024)

Train on 89403 samples, validate on 29802 samples
Epoch 1/1
89403/89403 [==============================] - 874s 10ms/step - loss: 1.4679 - acc: 0.9089 - val_loss: 1.4494 - val_acc: 0.9101


In [54]:
score = model.evaluate(XX_test, y_test, batch_size=528)
score

 1056/17029 [>.............................] - ETA: 1:15

KeyboardInterrupt: 

In [ ]:
# plot loss vs # model epochs trained
plt.plot(emb.history['val_loss'])
plt.plot(emb.history['loss'])
plt.grid()
plt.xlabel('epoch')
plt.ylabel('loss')
plt.legend(['valid','train'])

In [ ]:
# plot accuracy vs # epochs
plt.plot(emb.history['val_acc'])
plt.plot(emb.history['acc'])
plt.legend(['valid','train'])
plt.grid()
plt.xlabel('epoch')
plt.ylabel('accuracy')

In [ ]:
from sklearn.feature_selection import SelectKBest, f_classif, chi2
selector1 = SelectKBest(f_classif)
selector2 = SelectKBest(chi2)
top_features1 = selector1.fit_transform(X, y)
top_features2 = selector2.fit_transform(X, y)
mask1 = selector1.get_support()
mask2 = selector2.get_support()
columns1 = []
columns2 = []

In [ ]:
for col in range(X.shape[1]):
    if mask1[col] == True:
        columns1.append(df_a.columns[col])
    if mask2[col] == True:
        columns2.append(df_a.columns[col])
        
sc1 = pd.DataFrame(selector1.scores_, columns=["anova"], index=features)
sc1 = sc1.sort_values(by=["anova"], ascending=False)
sc2 = pd.DataFrame(selector2.scores_, columns=["chi2"], index=features)
sc2 = sc2.sort_values(by=["chi2"], ascending=False)

sc2

In [316]:
sc1

anova
recoveries                  25573.532963
collection_recovery_fee     20819.809346
out_prncp                    3899.481568
out_prncp_inv                3899.344939
sub_grade                    2826.455061
grade                        2692.279105
int_rate                     2364.082435
last_credit_pull_d           2280.071242
total_rec_prncp              2128.495969
last_pymnt_amnt              1402.007953
total_rec_late_fee           1311.955889
total_pymnt_inv               730.686187
total_pymnt                   730.622816
term                          503.851080
initial_list_status           435.041067
inq_last_6mths                395.748776
percent_bc_gt_75              315.234467
verification_status           311.893940
bc_open_to_buy                294.147769
tot_hi_cred_lim               265.080930
bc_util                       258.521159
acc_open_past_24mths          254.336035
total_bc_limit                243.719122
num_tl_op_past_12m            232.833345
avg_cur_bal                   227.955723
revol_util                    220.689589
tot_cur_bal                   209.486297
total_rev_hi_lim              191.133637
home_ownership                183.869601
mo_sin_rcnt_tl                155.792559
...                                  ...
num_rev_accts                  10.436831
num_op_rev_tl                   9.102186
dti                             9.009622
open_acc                        5.429047
num_sats                        5.387397
purpose                         4.647393
num_tl_90g_dpd_24m              4.513916
delinq_2yrs                     4.256683
num_il_tl                       4.028025
num_accts_ever_120_pd           3.230501
total_acc                       3.190037
issue_d                         2.748080
collections_12_mths_ex_med      2.729008
tax_liens                       2.517534
zip_code                        2.419592
pct_tl_nvr_dlq                  2.130088
pub_rec_bankruptcies            1.142671
tot_coll_amt                    1.110790
chargeoff_within_12_mths        0.592710
acc_now_delinq                  0.486194
num_tl_120dpd_2m                0.483000
pub_rec                         0.457909
addr_state                      0.255222
earliest_cr_line                0.104882
num_bc_sats                     0.076557
num_tl_30dpd                    0.030680
delinq_amnt                     0.002827
pymnt_plan                           NaN
policy_code                          NaN
hardship_flag                        NaN

[84 rows x 1 columns]

In [58]:
# classical ML methods
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()

from sklearn import svm
svm = svm.SVC()

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()

from sklearn.neural_network import MLPClassifier
#mlp = MLPClassifier(solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1)
mlp = MLPClassifier(solver='sgd', random_state=1)

In [56]:
# finetuning parameters
from sklearn.grid_search import GridSearchCV, RandomizedSearchCV
from sklearn import metrics
from scipy.stats import randint as sp_randint

/home/spaceape/anaconda3/envs/keras/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/home/spaceape/anaconda3/envs/keras/lib/python3.6/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


In [41]:
#import scipy
# SVM
# [1, 10, 30, 50, 70, 80, 90, 100, 120, 150, 1000]
# parameters_svm = {"kernel" : ["linear", "rbf"],
#                   "C" : [1, 10, 100, 1000],
#                   "gamma" : [0, 0.001]}

parameters_svm = {"kernel" : ["linear", "rbf"],
                  "C" : [1, 10, 100, 1000],
                  "gamma" : [0.001]}

# parameters_svm = {'C': scipy.stats.expon(scale=100), 'gamma': scipy.stats.expon(scale=.1),
#                   'kernel': ['rbf'], 'class_weight':['balanced', None]}
# svm_opt = RandomizedSearchCV(svm, parameters_svm, 1)
svm_opt = GridSearchCV(svm, parameters_svm,refit=True)

In [ ]:
svm_opt.fit(X_train, y_train)
print ("SVM parameters: ", svm_opt.best_params_)
print ("SVM accuracy: ",  svm_opt.best_estimator_.score(X_test, y_test))
#print ("SVM best estimator: ",  svma_opt.best_estimator_)
print ("SVM precision score: ",  metrics.precision_score(y_test, svm_opt.predict(X_test)))
print ("SVM recall score: ",  metrics.recall_score(y_test, svm_opt.predict(X_test)))

In [220]:
parameters_rf = {"max_depth": [3, 20],
              "max_features": sp_randint(1, 11),
              "min_samples_split": sp_randint(2, 15),
              "min_samples_leaf": sp_randint(1, 15),
              "n_estimators": sp_randint(10, 36),
              "bootstrap": [True],
              "criterion": ["gini", "entropy"]}
rf_opt = RandomizedSearchCV(rf, parameters_rf, 50)
rf_opt.fit(X_train, y_train)
print ("decision tree parameters: ", rf_opt.best_params_)
print ("decision tree accuracy: ",  rf_opt.best_estimator_.score(X_test, y_test))
#print ("decision tree best estimator: ",  dt_opt.best_estimator_)
print ("decision tree precision score: ",  metrics.precision_score(y_test, rf_opt.predict(X_test)))
print ("decision tree recall score: ",  metrics.recall_score(y_test, rf_opt.predict(X_test)))
print ("decisiton tree F1 score: ",  metrics.f1_score(y_test, rf_opt.predict(X_test)))

decision tree parameters:  {'bootstrap': True, 'criterion': 'entropy', 'max_depth': 20, 'max_features': 8, 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 33}
decision tree accuracy:  0.9962401597932088
decision tree precision score:  1.0
decision tree recall score:  0.9580602883355177


In [222]:
lr.fit(XX_train, y_train)
print ("logistic regression accuracy: ", lr.score(XX_test, y_test))
print ("logistic regression precision score: ",  metrics.precision_score(y_test, lr.predict(XX_test)))
print ("logistic regression recall score: ",  metrics.recall_score(y_test, lr.predict(XX_test)))
print ("logistic regression F1 score: ",  metrics.f1_score(y_test, lr.predict(XX_test)))

logistic regression accuracy:  0.9956526847608976
logistic regression precision score:  0.9959016393442623
logistic regression recall score:  0.9554390563564875
logistic regression F1 score:  0.9752508361204013


In [ ]:
parameters_mlp = {"alpha": [1e-4, 1e-5, 1e-6],
                 "hidden_layer_sizes": [(5,5), (10,5), (15,5), (20,), (25,), (15,), (20,), (25,)]}
parameters={
'learning_rate': ["constant", "invscaling", "adaptive"],
"hidden_layer_sizes": [(64, 15), (64, )],
'alpha': [1e-4, 1e-5, 1e-6],
'activation': ["logistic", "relu", "Tanh"]
}

mlp_opt = GridSearchCV(estimator=MLPClassifier(),param_grid=parameters,verbose=2,cv=10)
#mlp_opt = RandomizedSearchCV(mlp, parameters_mlp, n_iter=50, cv=10)

In [ ]:
mlp_opt = mlp_opt.fit(X_train, y_train)
print(mlp_opt.best_params_)
pred = mlp_opt.predict(X_test)

print ("confusion matrix: ", "\n", metrics.confusion_matrix(y_test, pred))
print ("accuracy score: ", metrics.accuracy_score(y_test, pred))
print ("precision score: ", metrics.precision_score(y_test, pred))
print ("recall score: ", metrics.recall_score(y_test, pred))
print ("F1 score: ", metrics.f1_score(y_test, pred))

In [ ]:
# METRICS

# ROC curve
fpr_lr, tpr_lr, _ = metrics.roc_curve(y_test, lr.predict_proba(X_test)[:,1])
fpr_rf, tpr_rf, _ = metrics.roc_curve(y_test, rf_opt.predict_proba(X_test)[:,1])
#fpr_svm, tpr_svm, _ = metrics.roc_curve(y_test, svm.predict(X_test))
fpr_mlp, tpr_mlp, _ = metrics.roc_curve(y_test, mlp.predict_proba(X_test)[:,1])


plt.figure(figsize=(10,6))
plt.plot(fpr_lr, tpr_lr, color='darkorange', label='Logistic Regression (area = %0.2f)' % metrics.auc(fpr_lr, tpr_lr))
#plt.plot(fpr_rf, tpr_rf, color='green', label='Logistic Regression (area = %0.2f)' % metrics.auc(fpr_rf, tpr_rf))
#plt.plot(fpr_mlp, tpr_mlp, color='blue', linestyle="-.", label='MLP (area = %0.2f)' % metrics.auc(fpr_mlp, tpr_mlp))
plt.plot([0, 1], [0, 1], color='black', linestyle='--')
# plt.xlim([0.0, 1.0])
# plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('Receiver operating characteristic example')
plt.legend(loc="lower right")
#plt.savefig("ROC.eps")
plt.show()

In [ ]:
# parameters tuning using the simple train_test_split
# from sklearn.cross_validation import train_test_split
# features_train, features_test, labels_train, labels_test = \
#     train_test_split(features, labels, test_size=0.3, random_state=42)

One of the most important step is feature selection, as usually the best statistical machine learning models deals with it very effectively and not only select the best features but also create them from availiable variables. The more complex approach to feature selection is often refered as feature engineering which beside feature selection include creating a new features. People working on these models may utilitize their expertise from particular field in order to get more signal bearing features and thus better model.  
Regarding feature selection the lending club dataset has too high dimensionality as it contains a 137?? features, which is from modelling point of a view not applicable. The goal is to have model based on as few as possible features but with very high prediction and generalization ability. Thus is neccessary to execute so called feature selection - procces which helps to achieve those goals mentioned in previous sentence and also it usually improves estimators accuracy as well. 

describe at least one univariate and one multivariate feature selection method.

Popsat take obecne preprocessing - the majority of time during building of model is consumed by preprocessing steps - including data cleaning, data wrangling, etc 

(2004) evaluating...
The actual
mining or extraction of patterns from the data
requires the data to be clean since input data are
the primary, if not the only, source of knowledge
in these systems. Cleaning and preprocessing data
involves several steps including procedures for
handling incomplete, noisy, or missing data; sam-
pling of appropriate data; feature selection; feature
construction; and also formatting the data as per
the representational requirements of methods (e.g.,
decision trees, neural networks) used to extract
knowledge from these data.

conclusion describe
generally good performing nonlinear method